lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -1220,6 +1269,220 @@ class CleanupFuncOps : public FuncOpPartialLoweringPattern {
+  using mlir::OpRewritePattern<calyx::CombGroupOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(calyx::CombGroupOp combGroupOp,
+                                PatternRewriter &rewriter) const override {
+    auto control =
+        combGroupOp->getParentOfType<calyx::ComponentOp>().getControlOp();
+    if (!SymbolTable::symbolKnownUseEmpty(combGroupOp.sym_nameAttr(), control))
+      return failure();
+
+    rewriter.eraseOp(combGroupOp);
+    return success();
+  }
+};
+
+/// GroupDoneOp's are terminator operations and should therefore be the last
+/// operator in a group. During group construction, we always append assignments
+/// to the end of a group, resulting in group_done ops migrating away from the
+/// terminator position. This pattern moves such ops to the end of their group.
+struct NonTerminatingGroupDonePattern
+    : mlir::OpRewritePattern<calyx::GroupDoneOp> {
+  using mlir::OpRewritePattern<calyx::GroupDoneOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(calyx::GroupDoneOp groupDoneOp,
+                                PatternRewriter & /*rewriter*/) const override {
+    Block *block = groupDoneOp->getBlock();
+    if (&block->back() == groupDoneOp)
+      return failure();
+
+    groupDoneOp->moveBefore(groupDoneOp->getBlock(),
+                            groupDoneOp->getBlock()->end());
+    return success();
+  }
+};
+
+/// When building groups which contain accesses to multiple sequential
+/// components, a group_done op is created for each of these. This pattern
+/// and's each of the group_done values into a single group_done.
+struct MultipleGroupDonePattern : mlir::OpRewritePattern<calyx::GroupOp> {
+  using mlir::OpRewritePattern<calyx::GroupOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(calyx::GroupOp groupOp,
+                                PatternRewriter &rewriter) const override {
+    auto groupDoneOps = SmallVector<calyx::GroupDoneOp>(
+        groupOp.getBody()->getOps<calyx::GroupDoneOp>());
+
+    if (groupDoneOps.size() <= 1)
+      return failure();
+
+    /// Create an and-tree of all calyx::GroupDoneOp's.
+    rewriter.setInsertionPointToEnd(groupDoneOps[0]->getBlock());
+    Value acc = groupDoneOps[0].src();
+    for (auto groupDoneOp : llvm::makeArrayRef(groupDoneOps).drop_front(1)) {
+      auto newAndOp = rewriter.create<comb::AndOp>(groupDoneOp.getLoc(), acc,
+                                                   groupDoneOp.src());
+      acc = newAndOp.getResult();
+    }
+
+    /// Create a group done op with the complex expression as a guard.
+    rewriter.create<calyx::GroupDoneOp>(
+        groupOp.getLoc(),
+        rewriter.create<hw::ConstantOp>(groupOp.getLoc(), APInt(1, 1)), acc);
+    for (auto groupDoneOp : groupDoneOps)
+      rewriter.eraseOp(groupDoneOp);
+
+    return success();
+  }
+};
+
+/// Returns true if 'child' is within the child tree of 'parent'.
+template <typename TParent>
+static bool isChildOp(TParent &parent, Operation *child) {
```

Comment:
Nit: if Parent is a sub-class of Operation*, just pass by value.

Reply:
Unsure where this will be used, but can the body just be replaced with `getParentOfType<TParent>()`, and seeing if that is equivalent to the `parent`?

Reply:
Unused, and should have been removed - thank you for catching it.